### PR TITLE
To memory review pass

### PR DESCRIPTION
**Pull request by ivirshup**
A bit of a belated review on the `.to_memory` method. Main changes:

* Renamed and simplified internal function a bit (`_create_anndata` -> `_mutated_copy`)
* Made `to_memory` work with `raw`
* Made `to_memory` work with dense backed arrays
* More tests
* Centralized subsetting of backed arrays (added `h5py.Dataset` and `SparseDataset` methods to `_subset`)
* Fixed bug in `assert_equals` where if the first argument didn't have raw, the second arguments `raw` wasn't checked

Could probably use a benchmark run.

TODO

- [x] Probably make `.to_memory` error if the object isn't backed.
- [x] Change log

### CODE REVIEW

**Comment by ivirshup** [comment]
@VolkerBergen two quick questions about this:

### `raw`

Originally this didn't bring `.raw` from a backed object into memory. Was this a design decision or a technical limitation?

If it was about not wanting to read too much data into memory, `raw` can be `del`'d from the backed object before bringing it into memory.

### Warning vs error

Originally this warned if `to_memory` was called on a in memory object. I took a bit more conservative approach and made it error. My philosophy here is that if you plan on handling backed objects, and they need special handling, it should be done explicitly.

Does this sound reasonable?

**Review by VolkerBergen — APPROVED** [review]
`.raw` was more of a technical limitation (quite some effort to implement). Loading `raw` in memory would be useful when subsetting along `.obs`, e.g., when reading a small celltype from a huge atlas. In our case, `.raw` was not relevant, thus didn't follow that up. 

The error makes sense, of course. 

I have no objections with the implementation. Thanks, looks great!

File: anndata/_core/anndata.py [hunk 1]
```diff
@@ -612,18 +612,8 @@ def X(self) -> Optional[Union[np.ndarray, sparse.spmatrix, ArrayView]]:
             # TODO: This should get replaced/ handled elsewhere
```
Comment (ivirshup) [inline anndata/_core/anndata.py:612]:
```suggestion
```

**Comment by ivirshup** [comment]
Great, thanks for the review!

**Pull request merged**